File: algorithms/medium/minimum_number_of_operations_to_reinitialize_a_permutation.py
class Solution:
    def reinitializePermutation(self, n: int) -> int:
        perm = [i for i in range(n)]
        arr = perm[:]
        steps = 0
        while True:
            arr = [arr[i//2] if i % 2 == 0 else arr[n//2 + (i-1)//2] for i in range(n)]
            steps += 1
            if arr == perm:
                return steps

# Main section
for n in [
               2,
               4,
               6,
               8,
              10,
              72,
             100,
             598,
             900,
            1000,
         ]:
    print(f'n = {n}')
    sol = Solution()
    r = sol.reinitializePermutation(n)
    print(f'r = {r}')
    print('================')

File: algorithms/medium/minimum_number_of_operations_to_reinitialize_a_permutation.go
package main

import "fmt"

func reinitializePermutation(n int) int {
	var perm, arr []int
	for i := 0; i < n; i++ {
		perm = append(perm, i)
		arr = append(arr, i)
	}
	steps := 0
	for true {
		tmp := make([]int, n)
		for i := 0; i < n; i++ {
			if i%2 == 0 {
				tmp[i] = arr[i/2]
			} else {
				tmp[i] = arr[n/2+(i-1)/2]
			}
		}
		copy(arr, tmp)
		steps++
		if areEqual(arr, perm) {
			return steps
		}
	}
	return 0
}

func areEqual(arr1, arr2 []int) bool {
	N, M := len(arr1), len(arr2)
	if N != M {
		return false
	}
	for i := 0; i < N; i++ {
		if arr1[i] != arr2[i] {
			return false
		}
	}
	return true
}

func main() {
	data := []int{
                       2,
                       4,
                       6,
                       8,
                      10,
                      72,
                     100,
                     598,
                     900,
                    1000,
	        }
	for _, n := range data {
		fmt.Printf("n = %d\n", n)
		r := reinitializePermutation(n)
		fmt.Printf("r = %d\n", r)
		fmt.Println("========================")
	}
}

